syntax = "proto3";

package arvos;

// Service definition for bidirectional sensor streaming
service SensorStream {
    // Bidirectional streaming - iPhone sends sensor data, server can send control commands
    rpc StreamSensors(stream SensorMessage) returns (stream ControlMessage);
}

// Main message wrapper - contains one of many sensor types
message SensorMessage {
    uint64 timestamp_ns = 1;
    
    oneof data {
        HandshakeData handshake = 10;
        IMUData imu = 11;
        GPSData gps = 12;
        PoseData pose = 13;
        CameraFrameMetadata camera_metadata = 14;
        DepthFrameMetadata depth_metadata = 15;
        WatchIMUData watch_imu = 16;
        WatchAttitudeData watch_attitude = 17;
        WatchMotionActivityData watch_activity = 18;
        StatusData status = 19;
        ErrorData error = 20;
    }
    
    // Optional: Binary data follows (for camera/depth frames)
    bytes binary_data = 100;
}

// Control messages from server to iPhone
message ControlMessage {
    uint64 timestamp_ns = 1;
    
    oneof command {
        StartRecordingCommand start_recording = 10;
        StopRecordingCommand stop_recording = 11;
        ChangeModeCommand change_mode = 12;
        AcknowledgmentCommand ack = 13;
    }
}

// MARK: - Handshake & Device Info

message HandshakeData {
    string device_name = 1;
    string device_model = 2;
    string os_version = 3;
    string app_version = 4;
    DeviceCapabilities capabilities = 5;
}

message DeviceCapabilities {
    bool has_lidar = 1;
    bool has_arkit = 2;
    bool has_gps = 3;
    bool has_imu = 4;
    bool has_watch = 5;
    repeated string supported_modes = 6;
}

// MARK: - Sensor Data Types

message IMUData {
    Vector3 angular_velocity = 1;     // rad/s
    Vector3 linear_acceleration = 2;   // m/s²
    Vector3 magnetic_field = 3;        // μT (optional)
    Vector3 gravity = 4;                // m/s² (optional)
    Vector3 attitude = 5;               // roll, pitch, yaw in radians (optional)
}

message GPSData {
    double latitude = 1;                // degrees
    double longitude = 2;               // degrees
    double altitude = 3;                // meters
    double horizontal_accuracy = 4;     // meters
    double vertical_accuracy = 5;       // meters
    double speed = 6;                   // m/s
    double course = 7;                  // degrees
}

message PoseData {
    Vector3 position = 1;               // meters (x, y, z)
    Quaternion orientation = 2;         // quaternion (x, y, z, w)
    string tracking_state = 3;          // "normal", "limited_*", "not_available"
    Matrix4x4 transform = 4;            // optional 4x4 transform matrix
}

message CameraFrameMetadata {
    uint32 width = 1;
    uint32 height = 2;
    string format = 3;                  // "jpeg", "h264", "png"
    uint32 compressed_size = 4;
    CameraIntrinsics intrinsics = 5;    // optional
}

message DepthFrameMetadata {
    uint32 point_count = 1;
    float min_depth = 2;                // meters
    float max_depth = 3;                // meters
    string format = 4;                  // "ply", "raw_depth"
    uint32 data_size = 5;
}

// Apple Watch Sensor Data

message WatchIMUData {
    Vector3 angular_velocity = 1;       // rad/s
    Vector3 linear_acceleration = 2;    // m/s²
    Vector3 gravity = 3;                // m/s²
}

message WatchAttitudeData {
    Quaternion quaternion = 1;          // x, y, z, w
    float pitch = 2;                    // radians
    float roll = 3;                     // radians
    float yaw = 4;                      // radians
    string reference_frame = 5;         // "xArbitraryZVertical", "xArbitraryCorrectedZVertical", etc.
}

message WatchMotionActivityData {
    string state = 1;                   // "walking", "running", "cycling", "vehicle", "stationary", "unknown"
    float confidence = 2;               // 0.0 - 1.0
}

// Status and Error Messages

message StatusData {
    string status = 1;                  // "connected", "streaming", "recording", "stopped", "error"
    string message = 2;                 // optional message
    string session_id = 3;              // optional session ID
}

message ErrorData {
    string error = 1;
    string details = 2;                 // optional details
}

// MARK: - Control Commands

message StartRecordingCommand {
    string session_id = 1;              // optional session ID
    string mode = 2;                    // optional mode
}

message StopRecordingCommand {
    string session_id = 1;              // optional session ID
}

message ChangeModeCommand {
    string mode = 1;                    // "Mapping", "Full Sensor", etc.
}

message AcknowledgmentCommand {
    string message_id = 1;              // ID of acknowledged message
    bool success = 2;
}

// MARK: - Common Data Structures

message Vector3 {
    float x = 1;
    float y = 2;
    float z = 3;
}

message Quaternion {
    float x = 1;
    float y = 2;
    float z = 3;
    float w = 4;
}

message Matrix4x4 {
    repeated float values = 1;          // 16 values in column-major order
}

message CameraIntrinsics {
    float fx = 1;                       // focal length x
    float fy = 2;                       // focal length y
    float cx = 3;                       // principal point x
    float cy = 4;                       // principal point y
    repeated float distortion = 5;      // distortion coefficients (k1, k2, p1, p2, k3...)
}

